coderd/healthcheck/provisioner.go

```diff
@@ -0,0 +1,141 @@
+package healthcheck
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/mod/semver"
+
+	"github.com/coder/coder/v2/buildinfo"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/healthcheck/health"
+	"github.com/coder/coder/v2/coderd/provisionerdserver"
+	"github.com/coder/coder/v2/coderd/util/apiversion"
+	"github.com/coder/coder/v2/coderd/util/ptr"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+// @typescript-generate ProvisionerDaemonsReport
+type ProvisionerDaemonsReport struct {
+	Severity  health.Severity  `json:"severity"`
+	Warnings  []health.Message `json:"warnings"`
+	Dismissed bool             `json:"dismissed"`
+	Error     *string          `json:"error"`
+
+	ProvisionerDaemons []codersdk.ProvisionerDaemon `json:"provisioner_daemons"`
+}
+
+type ProvisionerDaemonsReportOptions struct {
+	CurrentVersion    string
+	CurrentAPIVersion *apiversion.APIVersion
+
+	// ProvisionerDaemonsFn is a function that returns ProvisionerDaemons.
+	// Satisfied by database.Store.ProvisionerDaemons
+	ProvisionerDaemonsFn func(context.Context) ([]database.ProvisionerDaemon, error)
+
+	TimeNowFn     func() time.Time
+	StaleInterval time.Duration
+
+	Dismissed bool
+}
+
+func (r *ProvisionerDaemonsReport) Run(ctx context.Context, opts *ProvisionerDaemonsReportOptions) {
+	r.ProvisionerDaemons = make([]codersdk.ProvisionerDaemon, 0)
+	r.Severity = health.SeverityOK
+	r.Warnings = make([]health.Message, 0)
+	r.Dismissed = opts.Dismissed
+	now := opts.TimeNowFn()
+	if opts.StaleInterval == 0 {
+		opts.StaleInterval = provisionerdserver.DefaultHeartbeatInterval * 3
+	}
+
+	if opts.CurrentVersion == "" {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("Developer error: CurrentVersion is empty!")
+		return
+	}
+
+	if opts.CurrentAPIVersion == nil {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("Developer error: CurrentAPIVersion is nil!")
+		return
+	}
+
+	if opts.ProvisionerDaemonsFn == nil {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("Developer error: ProvisionerDaemonsFn is nil!")
+		return
+	}
+
+	// nolint: gocritic // need an actor to fetch provisioner daemons
+	daemons, err := opts.ProvisionerDaemonsFn(dbauthz.AsSystemRestricted(ctx))
+	if err != nil {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("error fetching provisioner daemons: " + err.Error())
+		return
+	}
+
+	for _, daemon := range daemons {
+		r.ProvisionerDaemons = append(r.ProvisionerDaemons, convertProvisionerDaemon(daemon))
+	}
+
+	if len(r.ProvisionerDaemons) == 0 {
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref("No provisioner daemons found!")
+		return
+	}
+
+	for _, daemon := range daemons {
+		// Daemon never connected, skip.
+		if !daemon.LastSeenAt.Valid {
+			continue
+		}
+		// Daemon has gone away, skip.
+		if now.Sub(daemon.LastSeenAt.Time) > (opts.StaleInterval) {
```

Comment:
Thinking out loud, is it possible that a daemon reports an error, but after that, the last seen isn't updated. Then after stale interval the error disappears, and perhaps nobody ever notices it?

Also, don't we want to apply the same rules to `r.ProvisionerDaemons`?

Reply:
> is it possible that a daemon reports an error, but after that, the last seen isn't updated. Then after stale interval the error disappears, and perhaps nobody ever notices it?

Yes, and this is by design. If a provisioner daemon connects at some point, has a transient error, and then never heartbeats, I don't think it makes sense to warn about this.

Reply:
That's fair. I was entertaining the possibility of some error state resulting in the provisioner exiting, crashing or plain stopping communication with server. But I didn't have anything concrete in mind, so this is fine.